### PR TITLE
[WIP] bpo-23689: re module, fix memory leak when a match is terminated by a signal

### PR DESCRIPTION
**Pull request by ghost**
**A draft**

Count the number of REPEAT when compiling a pattern, and allocate an array in `SRE_STATE`.
At any time, a REPEAT will have at most one in active, so a `SRE_REPEAT` array is fine.

https://bugs.python.org/issue23689

### CODE REVIEW

**Comment by animalize** [comment]
@serhiy-storchaka 
This is a draft. If use 5~7 days, it may achieve better quality. 
I can wait until your PR "bpo-47152: Convert the re module into a package" is merged, then I redo this PR.

File: Modules/sre_lib.h [hunk 1]
```diff
@@ -1032,16 +1033,13 @@ SRE(match)(SRE_STATE* state, const SRE_CODE* pattern, int toplevel)
         case SRE_OP_REPEAT:
             /* create repeat context.  all the hard work is done
                by the UNTIL operator (MAX_UNTIL, MIN_UNTIL) */
-            /* <REPEAT> <skip> <1=min> <2=max> item <UNTIL> tail */
+            /* <REPEAT> <skip> <1=min> <2=max> <3=repeat_index> item <UNTIL> tail */
             TRACE(("|%p|%p|REPEAT %d %d\n", ctx->pattern, ctx->ptr,
```
Comment (serhiy-storchaka) [inline Modules/sre_lib.h]: Add repeat_index.

File: Modules/sre_lib.h [hunk 2]
```diff
@@ -806,7 +807,7 @@ SRE(match)(SRE_STATE* state, const SRE_CODE* pattern, int toplevel)
                collecting backtracking points.  for other cases,
                use the MAX_REPEAT operator */
 
-            /* <REPEAT_ONE> <skip> <1=min> <2=max> item <SUCCESS> tail */
+            /* <REPEAT_ONE> <skip> <1=min> <2=max> <3=repeat_index> item <SUCCESS> tail */
```
Comment (serhiy-storchaka) [inline Modules/sre_lib.h]: Seems `repeat_index` is not used here.

**Review by serhiy-storchaka — COMMENTED** [review]
Seems `repeat_index` is only needed for `REPEAT`.

File: Modules/sre_lib.h (hunk 1 above)
Comment (ghost) [reply Modules/sre_lib.h]: Ok, I will.

File: Modules/sre_lib.h (hunk 2 above)
Comment (ghost) [reply Modules/sre_lib.h]: I removed `repeat_index` for these 3 OPs: `REPEAT_ONE`, `MIN_REPEAT_ONE`, `POSSESSIVE_REPEAT_ONE`.

**Comment by animalize** [comment]
This PR has not been carefully checked yet, it will take a few days to do this.

Just want to ask your arrangement, is this PR merged first, or your PR first?
I tend to merge your PR first, then I can carefully check when you  are preparing your PR.

**Comment by serhiy-storchaka** [comment]
I prefer to wait until this PR is merged. It is easy to redo my changes.

File: Modules/sre_lib.h [hunk 3]
```diff
@@ -1200,10 +1200,10 @@ SRE(match)(SRE_STATE* state, const SRE_CODE* pattern, int toplevel)
 
         case SRE_OP_POSSESSIVE_REPEAT:
             /* create possessive repeat contexts. */
-            /* <POSSESSIVE_REPEAT> <skip> <1=min> <2=max> pattern
```
Comment (serhiy-storchaka) [inline Modules/sre_lib.h:1203]: What about POSSESSIVE_REPEAT? `repeat_index` is not used here. MAX_UNTIL and MIN_UNTIL are not used with POSSESSIVE_REPEAT, so there is no problem with keeping POSSESSIVE_REPEAT intact.

**Comment by animalize** [comment]
Ok, I will finish this PR in a few days.

File: Lib/sre_compile.py [hunk 4]
```diff
@@ -587,6 +591,9 @@ def _compile_info(code, pattern, flags):
         emit(MAXCODE)
         prefix = prefix[:MAXCODE]
     emit(min(hi, MAXCODE))
+    # REPEAT count
+    assert len(code) == _REPEAT_COUNT_OFFSET
+    emit(0)
```
Comment (serhiy-storchaka) [inline Lib/sre_compile.py:596]: It seems to me that INFO is only used for optimization. If you ignore it, all will work, but maybe slower.

Essential information like the number of groups etc is passed to `_sre.compile()` as separate arguments. The REPEAT count is of such kind. If it is not known we cannot start matching.

I propose to pass it as a separate argument. And in `_validate_inner()` check that all repeat_index are less than that count (like we validate group indices and  the MARK argument).

**Review by serhiy-storchaka — COMMENTED** [review]
What I propose:
1. Remove repeat_index in POSSESSIVE_REPEAT.
2. Remove REPEAT count in INFO and pass it as a separate argument to `_sre.compile()`.
3. Validate repeat_index for REPEAT in `_validate_inner()`.

File: Lib/sre_compile.py (hunk 4 above)
Comment (ghost) [reply Lib/sre_compile.py:596]: Sounds reasonable, I'll do these.
Please merge your PR first.

File: Modules/sre_lib.h (hunk 3 above)
Comment (ghost) [reply Modules/sre_lib.h:1203]: If I add more conditional checks, I'm afraid to slow down the pattern compilation.

Comment (serhiy-storchaka) [reply Modules/sre_lib.h:1203]: Not much. Checking `if op is POSSESSIVE_REPEAT` is cheap in comparison with parsing or optimization.

**Comment by animalize** [comment]
`CompileData.diff` changes `code` to a `CompileData` object, which may save intermediate states during compilation. **Only** apply this patch to `main` branch, MSVC2022 non-PGO release build runs `test_re.py` is 2.1% slower (0.283 sec -> 0.277 sec).

Another way is to temporarily store it in a field in `SRE_OP_INFO`, but this would make `SRE_OP_INFO` take up an unnecessary field.

To be honest, the whole patch is intricacy, so reviewing is harder than making it from zero. If you think so too, you may make a PR instead of this PR.

[CompileData.diff.txt](https://github.com/python/cpython/files/8385960/CompileData.diff.txt)  (.diff attachment is not supportted by GitHub, so I added .txt)

**Comment by animalize** [comment]
This simpler patch is faster (2.80 sec)
[simple_patch.diff.txt](https://github.com/python/cpython/files/8386016/simple_patch.diff.txt)

**Comment by animalize** [comment]
I can divide the patch into some commits, each commit only modify one simple step, which makes the review much easier. Then this PR will be closed.
I suggest merge your PR first.

**Pull request closed without merging**